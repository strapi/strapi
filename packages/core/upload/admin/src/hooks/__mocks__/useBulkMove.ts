export const useBulkMove = jest.fn().mockReturnValue({
  isLoading: false,
  error: null,
  move: jest.fn(),
});
