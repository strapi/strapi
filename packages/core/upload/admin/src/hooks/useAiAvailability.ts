import { useAIAvailability as useGlobalAIAvailability } from '@strapi/admin/strapi-admin/ee';

import { useSettings } from './useSettings';

export const useAIAvailability = () => {
  const isAiAvailable = useGlobalAIAvailability();
  const { status, data } = useSettings(isAiAvailable);

  if (!isAiAvailable) {
    return { status: 'success' as const, isEnabled: false };
  }

  return { status, isEnabled: data?.aiMetadata };
};
