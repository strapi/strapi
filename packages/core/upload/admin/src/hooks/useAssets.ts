import * as React from 'react';

import { useNotification, useFetchClient } from '@strapi/admin/strapi-admin';
import { useNotifyAT } from '@strapi/design-system';
import { useIntl } from 'react-intl';
import { useQuery } from 'react-query';

import { Query, GetFiles } from '../../../shared/contracts/files';
import { pluginId } from '../pluginId';

interface UseAssetsOptions {
  skipWhen?: boolean;
  query?: Query;
}

export const useAssets = ({ skipWhen = false, query = {} }: UseAssetsOptions = {}) => {
  const { formatMessage } = useIntl();
  const { toggleNotification } = useNotification();
  const { notifyStatus } = useNotifyAT();
  const { get } = useFetchClient();
  const { folderPath, _q, ...paramsExceptFolderAndQ } = query;

  let params: Query;

  if (_q) {
    params = {
      ...paramsExceptFolderAndQ,
      _q: encodeURIComponent(_q),
    };
  } else {
    params = {
      ...paramsExceptFolderAndQ,
      filters: {
        $and: [
          ...(paramsExceptFolderAndQ?.filters?.$and ?? []),
          {
            folderPath: { $eq: folderPath ?? '/' },
          },
        ],
      },
    };
  }

  const { data, error, isLoading } = useQuery<
    GetFiles.Response['data'],
    GetFiles.Response['error']
  >(
    [pluginId, 'assets', params],
    async () => {
      const { data } = await get('/upload/files', { params });

      return data;
    },
    {
      enabled: !skipWhen,
      staleTime: 0,
      cacheTime: 0,
      select(data) {
        if (data?.results && Array.isArray(data.results)) {
          return {
            ...data,
            results: data.results
              /**
               * Filter out assets that don't have a name.
               * So we don't try to render them as assets
               * and get errors.
               */
              .filter((asset) => asset.name)
              .map((asset) => ({
                ...asset,
                /**
                 * Mime and ext cannot be null in the front-end because
                 * we expect them to be strings and use the `includes` method.
                 */
                mime: asset.mime ?? '',
                ext: asset.ext ?? '',
              })),
          };
        }

        return data;
      },
    }
  );

  React.useEffect(() => {
    if (data) {
      notifyStatus(
        formatMessage({
          id: 'list.asset.at.finished',
          defaultMessage: 'The assets have finished loading.',
        })
      );
    }
  }, [data, formatMessage, notifyStatus]);

  React.useEffect(() => {
    if (error) {
      toggleNotification({
        type: 'danger',
        message: formatMessage({ id: 'notification.error' }),
      });
    }
  }, [error, formatMessage, toggleNotification]);

  return { data, error, isLoading };
};
