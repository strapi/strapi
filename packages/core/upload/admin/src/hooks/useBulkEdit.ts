import { useNotification, useFetchClient } from '@strapi/admin/strapi-admin';
import { useIntl } from 'react-intl';
import { useMutation, useQueryClient } from 'react-query';

import { BulkUpdateFiles } from '../../../shared/contracts/files';
import { pluginId } from '../pluginId';
import { getTrad } from '../utils';

interface FileInfoUpdate {
  name: string;
  alternativeText: string | null;
  caption: string | null;
  folder: number | null;
}

interface BulkEditParams {
  updates: Array<{
    id: number;
    fileInfo: FileInfoUpdate;
  }>;
}

export const useBulkEdit = () => {
  const { formatMessage } = useIntl();
  const { toggleNotification } = useNotification();
  const queryClient = useQueryClient();
  const { post } = useFetchClient();

  const bulkEditQuery = ({ updates }: BulkEditParams) => {
    return post('/upload/actions/bulk-update', { updates });
  };

  const mutation = useMutation<
    BulkUpdateFiles.Response,
    BulkUpdateFiles.Response['error'],
    BulkEditParams
  >(bulkEditQuery, {
    onSuccess(res) {
      const { data } = res;

      if (data && data.length > 0) {
        queryClient.refetchQueries([pluginId, 'assets'], { active: true });
        queryClient.refetchQueries([pluginId, 'asset-count'], { active: true });
        queryClient.refetchQueries([pluginId, 'folders'], { active: true });
      }

      toggleNotification({
        type: 'success',
        message: formatMessage({
          id: getTrad('modal.edit.success-label'),
          defaultMessage: 'Files have been successfully updated.',
        }),
      });
    },
  });

  const edit = (
    updates: Array<{
      id: number;
      fileInfo: FileInfoUpdate;
    }>
  ) => mutation.mutateAsync({ updates });

  return { ...mutation, edit };
};
