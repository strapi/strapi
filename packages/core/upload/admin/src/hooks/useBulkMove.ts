import { useNotification, useFetchClient } from '@strapi/admin/strapi-admin';
import { useIntl } from 'react-intl';
import { useMutation, useQueryClient } from 'react-query';

import { File, BulkMoveFiles } from '../../../shared/contracts/files';
import { Folder, BulkMoveFolders } from '../../../shared/contracts/folders';
import { pluginId } from '../pluginId';
import { getTrad } from '../utils';

export interface FolderWithType extends Folder {
  type: string;
}

export interface FileWithType extends File {
  type: string;
}

interface BulkMoveParams {
  destinationFolderId: number | string;
  filesAndFolders: Array<FolderWithType | FileWithType>;
}

// Define the shape of the accumulator object
type Payload = {
  fileIds?: number[];
  folderIds?: number[];
};

export const useBulkMove = () => {
  const { formatMessage } = useIntl();
  const { toggleNotification } = useNotification();
  const queryClient = useQueryClient();
  const { post } = useFetchClient();

  const bulkMoveQuery = ({ destinationFolderId, filesAndFolders }: BulkMoveParams) => {
    const payload = filesAndFolders.reduce<Payload>((acc, selected) => {
      const { id, type } = selected;
      const key = type === 'asset' ? 'fileIds' : 'folderIds';

      if (!acc[key]) {
        acc[key] = [];
      }

      acc[key]!.push(id);

      return acc;
    }, {});

    return post('/upload/actions/bulk-move', { ...payload, destinationFolderId });
  };

  const mutation = useMutation<
    BulkMoveFolders.Response | BulkMoveFiles.Response,
    BulkMoveFolders.Response['error'] | BulkMoveFiles.Response['error'],
    BulkMoveParams
  >(bulkMoveQuery, {
    onSuccess(res) {
      const {
        data: { data },
      } = res;

      if (data?.files?.length > 0) {
        queryClient.refetchQueries([pluginId, 'assets'], { active: true });
        queryClient.refetchQueries([pluginId, 'asset-count'], { active: true });
      }

      // folders need to be re-fetched in any case, because assets might have been
      // moved into a sub-folder and therefore the count needs to be updated
      queryClient.refetchQueries([pluginId, 'folders'], { active: true });

      toggleNotification({
        type: 'success',
        message: formatMessage({
          id: getTrad('modal.move.success-label'),
          defaultMessage: 'Elements have been moved successfully.',
        }),
      });
    },
  });

  const move = (
    destinationFolderId: number | string,
    filesAndFolders: Array<FolderWithType | FileWithType>
  ) => mutation.mutateAsync({ destinationFolderId, filesAndFolders });

  return { ...mutation, move };
};
