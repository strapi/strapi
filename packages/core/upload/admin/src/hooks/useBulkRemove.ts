import { useNotification, useFetchClient, adminApi } from '@strapi/admin/strapi-admin';
import { useIntl } from 'react-intl';
import { useMutation, useQueryClient } from 'react-query';
import { useDispatch } from 'react-redux';

import { BulkDeleteFiles, File } from '../../../shared/contracts/files';
import { pluginId } from '../pluginId';
import { getTrad } from '../utils';

import type { BulkDeleteFolders, FolderDefinition } from '../../../shared/contracts/folders';

export interface FileWithType extends File {
  type: string;
}

type BulkRemovePayload = Partial<BulkDeleteFiles.Request['body']> &
  Partial<BulkDeleteFolders.Request['body']>;

export const useBulkRemove = () => {
  const dispatch = useDispatch();
  const { toggleNotification } = useNotification();
  const { formatMessage } = useIntl();
  const queryClient = useQueryClient();
  const { post } = useFetchClient();

  const bulkRemoveQuery = (filesAndFolders: Array<FileWithType | FolderDefinition>) => {
    const payload = filesAndFolders.reduce<BulkRemovePayload>((acc, selected) => {
      const { id, type } = selected;
      const key = type === 'asset' ? 'fileIds' : 'folderIds';

      if (!acc[key]) {
        acc[key] = [];
      }

      acc[key]!.push(id);

      return acc;
    }, {});

    return post('/upload/actions/bulk-delete', payload);
  };

  const mutation = useMutation<
    BulkDeleteFiles.Response | BulkDeleteFolders.Response,
    BulkDeleteFiles.Response['error'] | BulkDeleteFolders.Response['error'],
    Array<FileWithType | FolderDefinition>
  >(bulkRemoveQuery, {
    onSuccess(res) {
      const {
        data: { data },
      } = res;

      if (data?.files?.length > 0) {
        queryClient.refetchQueries([pluginId, 'assets'], { active: true });
        queryClient.refetchQueries([pluginId, 'asset-count'], { active: true });
      }

      if (data?.folders?.length > 0) {
        queryClient.refetchQueries([pluginId, 'folders'], { active: true });
      }

      toggleNotification({
        type: 'success',
        message: formatMessage({
          id: getTrad('modal.remove.success-label'),
          defaultMessage: 'Elements have been successfully deleted.',
        }),
      });

      dispatch(adminApi.util.invalidateTags(['HomepageKeyStatistics']));
    },
    onError(error) {
      toggleNotification({ type: 'danger', message: error?.message });
    },
  });

  const remove = (...args: Parameters<typeof mutation.mutateAsync>) =>
    mutation.mutateAsync(...args);

  return { ...mutation, remove };
};
