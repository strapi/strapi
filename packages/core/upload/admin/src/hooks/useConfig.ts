import { useTracking, useNotification, useFetchClient } from '@strapi/admin/strapi-admin';
import { useIntl } from 'react-intl';
import { useMutation, useQuery, UseMutationResult, UseQueryResult } from 'react-query';

import { GetConfiguration, UpdateConfiguration } from '../../../shared/contracts/configuration';
import { pluginId } from '../pluginId';

const endpoint = `/${pluginId}/configuration`;
const queryKey = [pluginId, 'configuration'];

export const useConfig = () => {
  const { trackUsage } = useTracking();
  const { formatMessage } = useIntl();
  const { toggleNotification } = useNotification();
  const { get, put } = useFetchClient();

  const config: UseQueryResult<
    GetConfiguration.Response['data']['data'] | GetConfiguration.Response['error']
  > = useQuery(
    queryKey,
    async () => {
      const res: GetConfiguration.Response = await get(endpoint);

      return res.data.data;
    },
    {
      onError() {
        return toggleNotification({
          type: 'danger',
          message: formatMessage({ id: 'notification.error' }),
        });
      },
      /**
       * We're cementing that we always expect an object to be returned.
       */
      select: (data) => data || {},
    }
  );

  const putMutation: UseMutationResult<
    void,
    UpdateConfiguration.Response['error'],
    UpdateConfiguration.Request['body']
  > = useMutation(
    async (body) => {
      await put<UpdateConfiguration.Response>(endpoint, body);
    },
    {
      onSuccess() {
        trackUsage('didEditMediaLibraryConfig');
        config.refetch();
      },
      onError() {
        return toggleNotification({
          type: 'danger',
          message: formatMessage({ id: 'notification.error' }),
        });
      },
    }
  );

  return {
    config,
    mutateConfig: putMutation,
  };
};
