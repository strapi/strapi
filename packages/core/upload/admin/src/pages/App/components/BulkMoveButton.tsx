import * as React from 'react';

import { Button, Modal } from '@strapi/design-system';
import { Folder } from '@strapi/icons';
import { useIntl } from 'react-intl';

import { BulkMoveDialog } from '../../../components/BulkMoveDialog/BulkMoveDialog';

import type { File } from '../../../../../shared/contracts/files';
import type { Folder as FolderDefinition } from '../../../../../shared/contracts/folders';

interface FileWithType extends File {
  type: string;
}

interface FolderWithType extends FolderDefinition {
  type: string;
}

export interface BulkMoveButtonProps {
  onSuccess: () => void;
  currentFolder?: FolderWithType;
  selected: Array<FolderWithType | FileWithType>;
}

export const BulkMoveButton = ({ selected, onSuccess, currentFolder }: BulkMoveButtonProps) => {
  const { formatMessage } = useIntl();
  const [showConfirmDialog, setShowConfirmDialog] = React.useState(false);

  const handleConfirmMove = () => {
    setShowConfirmDialog(false);
    onSuccess();
  };

  return (
    <Modal.Root open={showConfirmDialog} onOpenChange={setShowConfirmDialog}>
      <Modal.Trigger>
        <Button variant="secondary" size="S" startIcon={<Folder />}>
          {formatMessage({ id: 'global.move', defaultMessage: 'Move' })}
        </Button>
      </Modal.Trigger>
      <BulkMoveDialog
        currentFolder={currentFolder}
        onClose={handleConfirmMove}
        selected={selected}
      />
    </Modal.Root>
  );
};
