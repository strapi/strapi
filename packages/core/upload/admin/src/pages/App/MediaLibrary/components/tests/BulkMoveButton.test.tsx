import { DesignSystemProvider } from '@strapi/design-system';
import { within, fireEvent, render, screen, waitFor } from '@testing-library/react';
import { IntlProvider } from 'react-intl';
import { QueryClient, QueryClientProvider } from 'react-query';
import { MemoryRouter } from 'react-router-dom';

import { useBulkMove } from '../../../../../hooks/useBulkMove';
import { BulkMoveButton } from '../BulkMoveButton';

import type { BulkMoveButtonProps } from '../BulkMoveButton';

jest.mock('../../../../../hooks/useBulkMove');
jest.mock('../../../../../hooks/useFolderStructure');

const FIXTURE_SELECTION = [
  {
    type: 'asset',
    createdAt: '2022-06-21T07:04:49.813Z',
    updatedAt: '2022-06-21T07:04:49.813Z',
    pathId: 1,
    path: '/1',
    files: { count: 0 },
    id: 1,
    children: { count: 0 },
    name: 'test',
  },
];

const setup = (
  props: BulkMoveButtonProps & {
    onClose?: () => void;
  } = {
    selected: [],
    onSuccess: jest.fn(),
  }
) => {
  const queryClient = new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        refetchOnWindowFocus: false,
      },
    },
  });

  return render(
    <QueryClientProvider client={queryClient}>
      <DesignSystemProvider>
        <MemoryRouter>
          <IntlProvider locale="en">
            <BulkMoveButton {...props} />
          </IntlProvider>
        </MemoryRouter>
      </DesignSystemProvider>
    </QueryClientProvider>
  );
};

describe('BulkMoveButton', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  test('renders', () => {
    const { container } = setup();

    expect(container).toMatchSnapshot();
  });

  test('opens destination dialog before the API call', async () => {
    const onSuccessSpy = jest.fn();
    const { getByText } = setup({
      onSuccess: onSuccessSpy,
      selected: FIXTURE_SELECTION as unknown as BulkMoveButtonProps['selected'],
    });
    const moveSpy = jest.fn().mockResolvedValueOnce({});

    (useBulkMove as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      error: null,
      move: moveSpy,
    });

    fireEvent.click(getByText('Move'));

    expect(getByText('Move elements to')).toBeInTheDocument();

    const dialog = screen.getByRole('dialog');
    const submit = within(dialog).getByRole('button', {
      name: /move/i,
    });
    fireEvent.click(submit);

    await waitFor(() => expect(moveSpy).toBeCalledWith('', FIXTURE_SELECTION));
    await waitFor(() => expect(onSuccessSpy).toBeCalled());
  });

  test('set default form values', () => {
    const { getByText } = setup({ onClose: jest.fn(), selected: [], onSuccess: jest.fn() });

    fireEvent.click(getByText('Move'));

    expect(screen.getByText('Media Library')).toBeInTheDocument();
  });

  test('set default form values with currentFolder', () => {
    const FIXTURE_PARENT_FOLDER = {
      id: 2,
      name: 'default folder name',
      updatedAt: '2022-06-21T15:35:36.932Z',
      createdAt: '2022-06-21T07:04:49.813Z',
      parent: null,
      path: '/2',
      pathId: 2,
    };
    const { getByText } = setup({
      currentFolder: FIXTURE_PARENT_FOLDER as unknown as BulkMoveButtonProps['currentFolder'],
      onClose: jest.fn(),
      selected: [],
      onSuccess: jest.fn(),
    });

    fireEvent.click(getByText('Move'));

    expect(screen.getByText(FIXTURE_PARENT_FOLDER.name)).toBeInTheDocument();
  });

  test('keeps destination dialog open if the API call errored', async () => {
    const FIXTURE_ERROR_MESSAGE = 'Failed to move folder';

    const onSuccessSpy = jest.fn();
    const { getByText } = setup({
      onSuccess: onSuccessSpy,
      selected: FIXTURE_SELECTION as unknown as BulkMoveButtonProps['selected'],
    });
    const moveSpy = jest.fn().mockRejectedValueOnce({
      response: {
        data: {
          error: {
            name: 'ValidationError',
            details: {
              errors: [
                {
                  path: [],
                  message: FIXTURE_ERROR_MESSAGE,
                  name: 'ValidationError',
                },
              ],
            },
          },
        },
      },
    });

    (useBulkMove as jest.Mock).mockReturnValueOnce({
      isLoading: false,
      error: null,
      move: moveSpy,
    });

    fireEvent.click(getByText('Move'));

    expect(getByText('Move elements to')).toBeInTheDocument();

    const dialog = screen.getByRole('dialog');
    const submit = within(dialog).getByRole('button', {
      name: /move/i,
    });
    fireEvent.click(submit);

    await waitFor(() => expect(moveSpy).toBeCalledWith('', FIXTURE_SELECTION));

    await waitFor(() => expect(onSuccessSpy).not.toBeCalled());
    expect(getByText('Move elements to')).toBeInTheDocument();
    expect(getByText(FIXTURE_ERROR_MESSAGE)).toBeInTheDocument();
  });
});
