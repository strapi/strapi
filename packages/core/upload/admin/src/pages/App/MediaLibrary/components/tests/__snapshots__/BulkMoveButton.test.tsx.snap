// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BulkMoveButton renders 1`] = `
.c0 {
  padding-inline-start: 16px;
  padding-inline-end: 16px;
  border-radius: 4px;
  cursor: pointer;
}

.c1 {
  gap: 8px;
  align-items: center;
  justify-content: center;
  flex-direction: row;
  display: inline-flex;
}

.c3 {
  align-items: center;
  flex-direction: row;
  display: flex;
}

.c4 {
  font-size: 1.2rem;
  line-height: 1.33;
  color: currentcolor;
  font-weight: 600;
}

.c5 {
  border: 0;
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c2 {
  height: 3.2rem;
  text-decoration: none;
  border: 1px solid #d9d8ff;
  background: #f0f0ff;
  color: #271fe0;
}

.c2:hover {
  background-color: #ffffff;
}

.c2:active {
  background-color: #ffffff;
  border: 1px solid #4945ff;
  color: #4945ff;
}

.c2[aria-disabled='true'] {
  border: 1px solid #dcdce4;
  background: #eaeaef;
  color: #666687;
  cursor: default;
}

@media (prefers-reduced-motion: no-preference) {
  .c2 {
    transition: background-color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),color 120ms cubic-bezier(0.25, 0.46, 0.45, 0.94),border-color 200ms cubic-bezier(0.25, 0.46, 0.45, 0.94);
  }
}

<div>
  <button
    aria-controls="radix-:r0:"
    aria-disabled="false"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="c0 c1 c2"
    data-state="closed"
    type="button"
  >
    <span
      aria-hidden="true"
      class="c3"
    >
      <svg
        fill="currentColor"
        height="16"
        viewBox="0 0 32 32"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M27 9H16.414L13 5.586A1.98 1.98 0 0 0 11.586 5H5a2 2 0 0 0-2 2v18.078A1.926 1.926 0 0 0 4.924 27H27.11A1.89 1.89 0 0 0 29 25.111V11a2 2 0 0 0-2-2M5 7h6.586l2 2H5z"
        />
      </svg>
    </span>
    <span
      class="c4"
    >
      Move
    </span>
  </button>
  <span
    class="c5"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </span>
</div>
`;
